app/instrumentation-tests/src/org/commcare/utils/DateTime.java

```diff
@@ -0,0 +1,361 @@
+/*
+ * Copyright (c) 2010 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.commcare.utils;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.GregorianCalendar;
+import java.util.TimeZone;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Immutable representation of a date with an optional time and an optional time zone based on <a
+ * href="http://tools.ietf.org/html/rfc3339">RFC 3339</a>.
+ *
+ * <p>
+ * Implementation is immutable and therefore thread-safe.
+ * </p>
+ *
+ * @since 1.0
+ * @author Yaniv Inbar
+ */
+public final class DateTime implements Serializable {
```

Comment:
I remember trying the Android DateTime methods for this but couldn't get them to work as expected. So I used this utility instead. It was maybe the `Z` offset that caused some trouble to me.

Reply:
Would you be able to have a relook and elaborate the issue you had with the `Z` offset.

Reply:
Putting it in here in case you're interested. The date string coming from HQ was like `2020-06-12T02:42:38.555065Z`
The closest I got to converting that to milliseconds was using `SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSsss'Z'")` and it gave `1591929785555` which is different than the expected `1591929758555`.
`java.time.Instant.parse()` gives correct result but it requires api level 26+ so can't use that. 
So, I'm using `org.joda.DateTime` right now.

Reply:
looks good.